### DISCUSSION TITLE
Texture array synchronization with GPU

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#237] [Q&A]
Short Version:
Is there any way to determine if a texture is in use by the GPU at runtime?

Long Version:
I'm using a large texture array as my game's entire texture system, where each new needed texture requests an open slot, and each texture being released frees up a slot. It uses reference counting to avoid duplicates and all of that, but I've encountered a pretty serious issue.

When a texture is removed, I swap it out with a blank pure white RGB 16x16 texture. And when an object needs to be refreshed, it is regenerated in a single frame, which causes the textures to be unlinked from the shader, then re-added almost immediately after (most likely to a different texture slot). I've noticed that there are random artifacts appearing on objects when this happens. The artifacts appear almost like tiny squares rather than polygons, which I'm assuming is part of my GPU's pixel renderer.  The artifacts appear either white or almost like a transparent-style grid. The grid looks almost like what you might see in Photoshop when you have a single layer that is transparent.

I don't know this for sure, but I believe this is happening because I'm swapping out textures while the GPU is still rendering with them. I'm assuming that since I'm not destroying the textures (just swapping them out), the GPU cannot use any type of reference counting to keep a lock on them. 

Is it possible to request this information at runtime, to see if a texture is in use by the GPU? Or perhaps a certain time within a frame that it would always be safe to swap out textures?

Appreciate any info!

**Comment by RobertMtx** [comment]
I ended up adding a simple 2-frame protection on registered textures, so they are reset only after being unregistered for 2 frames, which resolves the issue. The changes required adding additional systems to free the protected textures during actual transition events, like area changes, but it all worked out.

I would still be interested in hearing any advice related to this kind of thing. Thanks!

**Reply by TheMostDiligent** [reply]
There is no way to figure out if a specific resource is being currently used by GPU or not. What you can do is use fences to track the GPU execution progress. Fences are like signals - as long as the execution reaches a specific point, the fence is signaled.
Your double-frame approach should work.

Also, how do you bind/unbind the textures?

**Reply by RobertMtx** [reply]
Sorry, got wrapped up in development. I link textures to array elements in the shader using `Diligent::IShaderResourceVariable::SetArray()`. Usually one at a time, when the textures are requested for drawing or rendering. There are two arrays - one large array for 2D textures and one small array for cube maps. Basically, textures that are used every frame will stay linked, and others will be forced out to make room for new ones.

I have not used fences yet, but will definitely look into them. Thanks!

**Reply by TheMostDiligent** [reply]
If you just add new texture that has never been used before, there is no need to use fences.
Fences are needed if you are replacing a texture that was previously is use - you need fence to make sure that it is not in use by the GPU.
